File: src/main/kotlin/com/example/features/StringDemo.kt
package com.example.features

/**
 * Demonstrates full String function support including:
 * - Multi-line raw strings with trimIndent / trimMargin
 * - Case conversion (uppercase, lowercase)
 * - Whitespace handling (trim, trimStart, trimEnd, isBlank)
 * - Searching (indexOf, lastIndexOf, startsWith, endsWith, contains)
 * - Extraction (substring, split, lines)
 * - Transformation (replace, replaceFirst, reversed, repeat, padStart, padEnd)
 * - Prefix / suffix removal (removePrefix, removeSuffix)
 * - Numeric conversion (toInt, toDouble, toIntOrNull)
 * - Formatting (format)
 */

fun demoMultiLineString(): String {
    val poem = """
        Roses are red,
        Violets are blue,
        Kotlin is cool,
        And Lua is too!
    """.trimIndent()
    // Return the first line
    return poem.lines().first()
}

fun demoTrimMargin(): String {
    val text = """
        |Hello,
        |World!
    """.trimMargin()
    return text
}

fun demoStringCase(): String {
    val original = "Hello World"
    val upper = original.uppercase()
    val lower = original.lowercase()
    return "$upper/$lower"
}

fun demoStringTrim(): String {
    val s = "  hello  "
    return s.trim()
}

fun demoStringSearch(): String {
    val s = "Hello, Kotlin!"
    val idx = s.indexOf("Kotlin")
    val starts = s.startsWith("Hello")
    val ends = s.endsWith("!")
    val contains = s.contains("Kotlin")
    return "$idx/$starts/$ends/$contains"
}

fun demoStringSubstring(): String {
    val s = "Hello, World!"
    return s.substring(7, 12)
}

fun demoStringSplit(): Int {
    val csv = "apple,banana,cherry"
    val parts = csv.split(",")
    return parts.size
}

fun demoStringLines(): Int {
    val text = "line1\nline2\nline3"
    return text.lines().size
}

fun demoStringReplace(): String {
    val s = "Hello, World!"
    return s.replace("World", "Kotlin")
}

fun demoStringReplaceRegex(): String {
    val s = "foo123bar456"
    return s.replace(Regex("[0-9]+"), "#")
}

fun demoStringRepeat(): String {
    return "ab".repeat(3)
}

fun demoStringPad(): String {
    val n = "42"
    val padded = n.padStart(5, '0')
    val paddedEnd = n.padEnd(5, '-')
    return "$padded/$paddedEnd"
}

fun demoStringRemove(): String {
    val s = "prefixHello"
    val withoutPrefix = s.removePrefix("prefix")
    val withoutSuffix = "HelloSuffix".removeSuffix("Suffix")
    return "$withoutPrefix/$withoutSuffix"
}

fun demoStringConvert(): String {
    val n = "42"
    val i = n.toInt()
    val maybeNull = "abc".toIntOrNull()
    val notNull = "10".toIntOrNull()
    return "$i/${maybeNull == null}/${notNull != null}"
}

fun demoStringBlank(): String {
    val blank = "   "
    val notBlank = "hello"
    return "${blank.isBlank()}/${notBlank.isNotBlank()}"
}

fun demoStringFormat(): String {
    return "Hello, %s! You are %d years old.".format("World", 30)
}

fun demoStringReversed(): String {
    return "Hello".reversed()
}

fun demoStringContainsAndCount(): String {
    val s = "Hello, World!"
    val hasWorld = s.contains("World")
    val length = s.length
    return "$hasWorld/$length"
}

File: outputDir/vscripts/features/StringDemo.lua
-- package: com.example.features

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "features/StringDemo.kt", {["1-6"]=1,["7"]=17,["8-11"]=24,["12"]=28,["13-16"]=32,["17"]=36,["18"]=37,["19"]=38,["20-23"]=39,["24"]=43,["25-28"]=44,["29"]=48,["30"]=49,["31"]=50,["32"]=51,["33"]=52,["34-37"]=53,["38"]=57,["39-42"]=58,["43"]=62,["44"]=63,["45-48"]=64,["49"]=68,["50-53"]=69,["54"]=73,["55-58"]=74,["59"]=78,["60-63"]=79,["64-67"]=83,["68"]=87,["69"]=88,["70"]=89,["71-74"]=90,["75"]=94,["76"]=95,["77"]=96,["78-81"]=97,["82"]=101,["83"]=102,["84"]=103,["85"]=104,["86-89"]=105,["90"]=109,["91"]=110,["92-95"]=111,["96-99"]=115,["100-103"]=119,["104"]=123,["105"]=124,["106"]=125,["107-109"]=126})

function demoMultiLineString()
    local poem = ktox_trimIndent("\n        Roses are red,\n        Violets are blue,\n        Kotlin is cool,\n        And Lua is too!\n    ")
    return ktox_first(ktox_lines(poem))
end

function demoTrimMargin()
    local text = ktox_trimMargin("\n        |Hello,\n        |World!\n    ")
    return text
end

function demoStringCase()
    local original = "Hello World"
    local upper = ktox_uppercase(original)
    local lower = ktox_lowercase(original)
    return tostring(upper) .. "/" .. tostring(lower)
end

function demoStringTrim()
    local s = "  hello  "
    return ktox_trim(s)
end

function demoStringSearch()
    local s = "Hello, Kotlin!"
    local idx = ktox_indexOf(s, "Kotlin")
    local starts = ktox_startsWith(s, "Hello")
    local ends = ktox_endsWith(s, "!")
    local contains = ktox_contains(s, "Kotlin")
    return tostring(idx) .. "/" .. tostring(starts) .. "/" .. tostring(ends) .. "/" .. tostring(contains)
end

function demoStringSubstring()
    local s = "Hello, World!"
    return ktox_substring(s, 7, 12)
end

function demoStringSplit()
    local csv = "apple,banana,cherry"
    local parts = ktox_split(csv, ",")
    return #parts
end

function demoStringLines()
    local text = "line1" .. "\n" .. "line2" .. "\n" .. "line3"
    return #ktox_lines(text)
end

function demoStringReplace()
    local s = "Hello, World!"
    return ktox_replace(s, "World", "Kotlin")
end

function demoStringReplaceRegex()
    local s = "foo123bar456"
    return ktox_replaceRegex(s, "\[0-9\]+", "#")
end

function demoStringRepeat()
    return ktox_repeat("ab", 3)
end

function demoStringPad()
    local n = "42"
    local padded = ktox_padStart(n, 5, '0')
    local paddedEnd = ktox_padEnd(n, 5, '-')
    return tostring(padded) .. "/" .. tostring(paddedEnd)
end

function demoStringRemove()
    local s = "prefixHello"
    local withoutPrefix = ktox_removePrefix(s, "prefix")
    local withoutSuffix = ktox_removeSuffix("HelloSuffix", "Suffix")
    return tostring(withoutPrefix) .. "/" .. tostring(withoutSuffix)
end

function demoStringConvert()
    local n = "42"
    local i = ktox_toInt(n)
    local maybeNull = ktox_toIntOrNull("abc")
    local notNull = ktox_toIntOrNull("10")
    return tostring(i) .. "/" .. tostring(maybeNull == nil) .. "/" .. tostring(notNull ~= nil)
end

function demoStringBlank()
    local blank = "   "
    local notBlank = "hello"
    return tostring(ktox_isBlank(blank)) .. "/" .. tostring(ktox_isNotBlank(notBlank))
end

function demoStringFormat()
    return ktox_format("Hello, %s! You are %d years old.", "World", 30)
end

function demoStringReversed()
    return ktox_reversed("Hello")
end

function demoStringContainsAndCount()
    local s = "Hello, World!"
    local hasWorld = ktox_contains(s, "World")
    local length = #s
    return tostring(hasWorld) .. "/" .. tostring(length)
end

